src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Instance/InstanceDiscovery.cs

```diff
@@ -81,19 +82,19 @@ internal static class InstanceDiscovery
 
         private static SemaphoreSlim semaphore = new SemaphoreSlim(1, 1);
 
-        public static async Task<InstanceDiscoveryMetadataEntry> GetMetadataEntry(string host, bool validateAuthority,
+        public static async Task<InstanceDiscoveryMetadataEntry> GetMetadataEntry(Uri authority, bool validateAuthority,
             CallState callState)
         {
             InstanceDiscoveryMetadataEntry entry = null;
-            if (!InstanceCache.TryGetValue(host, out entry))
+            if (!InstanceCache.TryGetValue(authority.Host, out entry))
             {
                 await semaphore.WaitAsync().ConfigureAwait(false); // SemaphoreSlim.WaitAsync() will not block current thread
```

Comment:
I'm curious as to why you are using a [concurrent dictionary](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2?view=netframework-4.7.1), which is thread safe, _and_ using a semaphore, during which time a network call is occurring. I think you could just use the concurrent dictionary, since you are only reading and writing to the instance cache. It might be better to avoid performing time consuming operations (ex. a network call) while holding a lock.

Reply:
@jennyf19 Good question! I still remembered Brent and I spend quite some time discussing this tricky topic.

Long story short, the `ConcurrentDictionary` ensures its data can be safely accessed *via its methods* in multithread environment. In our case, the first `TryGetValue(...)` can safely try to get a value no matter what.

But `ConcurrentDictionary` does not and will not regulate how *other* code path prepare/consume those data. In our case, we want to make sure when the value does not exist, there will be one and only one network call occurring (more on this later). Such network call i.e. `DiscoverAsync(...)` happens outside of Dictionary methods, so we have to use a dedicate lock for it.

Also, in most other places, it is true that we won't want to hold a lock for a time consuming operations such as I/O calls, but in this case we happen to have a handy `SemaphoreSlim.WaitAsync()` which "locks" asynchronizely, so it does not actully block anything.

Lastly, ideally you would like to avoid such tricky requirement in the first place. Arguably, we could have relax the "only one network call at a time" requirement, and then such explicit lock won't be necessary. But that is a different topic anyway.

Reply:
Thanks for the explanation, @rayluo